src/storage/serde.rs

```diff
@@ -1,53 +1,57 @@
 use super::api::{Key, Value};
 use anyhow::{anyhow, Result};
+use derive_more::From;
 use std::fs::File;
 use std::io::{Read, Seek, SeekFrom, Write};
 use std::mem::size_of;
 
-fn write_key(k: &Key, w: &mut impl Write) -> Result<usize> {
-    let mut sz = w.write(&k.0.len().to_le_bytes())?;
-    sz += w.write(k.0.as_bytes())?;
-    Ok(sz)
+fn serialize_key(k: &Key, w: &mut impl Write) -> Result<usize> {
+    let pre_sz = w.write(&k.len().to_le_bytes())?;
+    let datum_sz = w.write(k.as_bytes())?;
+    Ok(pre_sz + datum_sz)
 }
 
-fn write_val(v: Option<&Value>, w: &mut impl Write) -> Result<usize> {
+fn serialize_val(v: &Value, w: &mut impl Write) -> Result<usize> {
     let sz = match v {
-        None => w.write(&(0usize).to_le_bytes())?,
-        Some(v) => match v {
-            Value::Bytes(bytes) => {
-                let mut sz = w.write(&bytes.len().to_le_bytes())?;
-                sz += w.write(bytes)?;
-                sz
-            }
-        },
+        Value::Tombstone => w.write(&(0usize).to_le_bytes())?,
+        Value::Bytes(bytes) => {
+            let pre_sz = w.write(&bytes.len().to_le_bytes())?;
+            let datum_sz = w.write(bytes)?;
+            pre_sz + datum_sz
+        }
     };
     Ok(sz)
 }
 
 /// This file format is applicable for both the commit log and ss tables.
-pub fn write_kv(k: &Key, v: Option<&Value>, file: &mut File) -> Result<usize> {
-    let mut sz = write_key(k, file)?;
-    sz += write_val(v, file)?;
-    Ok(sz)
+pub fn serialize_kv(k: &Key, v: &Value, w: &mut impl Write) -> Result<usize> {
+    let pre_sz = serialize_key(k, w)?;
+    let datum_sz = serialize_val(v, w)?;
+    Ok(pre_sz + datum_sz)
+}
+
+fn deserialize_key(bytes: Vec<u8>) -> Result<Key> {
+    Ok(Key(String::from_utf8(bytes)?))
+}
+
+fn deserialize_val(bytes: Vec<u8>) -> Value {
+    Value::Bytes(bytes)
 }
 
 enum FileItem {
     EOF,
-    Item(usize, Option<Vec<u8>>),
+    Skip(usize),
```

Comment:
perhaps a comment on what the different states represent. it's not clear to me what Skip and Empty denote.

Reply:
Oh oops I missed your comment. Let me do this.